Automattic/jetpack
# Jetpack Redirector: Add backend support for redirecting inbound links to their proper destination instead of 404ing

This is a Work in progress! Do not merge :)

See p1HpG7-58b-p2

---

#### Changes proposed in this Pull Request:

* Register a new `Jetpack_Redirect` Custom Post Type with no UI
* Attempt to resolve 404 requests to such a post type
  * Redirect on a hit
  * Do nothing on a miss
* Cache results to speed up subsequent requests

#### Testing instructions:



* Determine the post ID of the intended target (it can be any type of post including regular posts / pages, attachment pages, or custom post types). It will be referred to as `$target_post_id` below.
* Manually add a redirect (via `wp shell` or anywhere else you can run a one-off command):
```
// This can be pretty much any arbitrary path
$path_to_redirect = '/Your/way/KEWL/old/path/you/really/want/to/redirect';

wp_insert_post( array(
    'post_type' => 'jetpack-redirect',
    'post_status' => 'publish',
    'post_parent' => $target_post_id,
    'post_title' => $path_to_redirect,
    'post_name' => md5( $path_to_redirect ),
) );
```
* Attempt to fetch the path you used above.  For example:
`curl -I https://random-gibberish.jurassic.ninja/Your/way/KEWL/old/path/you/really/want/to/redirect`

You should see the path resolve to the intended content via a `301` response with custom header `x-jetpack-redirector: HIT`

It behave appropriately in all supported versions of PHP (~5.2+)


#### Proposed changelog entry for your changes:
